resolwe/flow/views/annotations.py

```diff
@@ -83,21 +79,45 @@ class AnnotationValueViewSet(
     queryset = AnnotationValue.objects.all()
     permission_classes = (get_permissions_class(),)
 
-    def get_serializer(self, *args, **kwargs):
-        """
-        Return the serializer instance that should be used for validating and
-        deserializing input, and for serializing output.
+    def _validate_values(self, annotation_values: Iterable[AnnotationValue]):
+        """Validate annotation values.
+
+        :raises ValidationError: if validation does not pass.
         """
-        kwargs["many"] = isinstance(self.request.data, list)
-        return super().get_serializer(*args, **kwargs)
+        validation_errors: list[ValidationError] = []
+        for value in annotation_values:
+            try:
+                value.validate()
+            except ValidationError as e:
+                validation_errors.append(e)
+        if validation_errors:
+            raise ValidationError(validation_errors)
+
+    def _bulk_create(self, annotation_values: Iterable[AnnotationValue]):
+        """Bulk create or update annotation values."""
+        also_update = self.request.method == "PUT"
+        inserted_values: list[AnnotationValue] = []
+        if also_update:
+            for annotation_value in annotation_values:
+                inserted_values.append(
+                    AnnotationValue.objects.update_or_create(
```

Comment:
Can you use `bulk_create` and set `update_conflicts` to `True`?

Reply:
I could, but then it would not return primary keys of inserted/updated objects.